drivers/gpio/gpio_shell.c

```diff
@@ -143,45 +137,74 @@ static int cmd_gpio_blink(const struct shell *sh,
 	uint8_t value = 0U;
 	size_t count = 0;
 	char data;
+	int rc = 0;
 
-	if (isdigit((unsigned char)argv[args_indx.index][0]) != 0) {
-		index = (uint8_t)atoi(argv[args_indx.index]);
-	} else {
+	dev = device_get_binding(argv[1]);
+	if (dev == NULL) {
+		shell_error(sh, "GPIO device %s not available", argv[1]);
+		return -ENODEV;
+	}
+
+	index = shell_strtoul(argv[2], 10, &rc);
+	if (rc < 0) {
 		shell_error(sh, "Wrong parameters for blink");
-		return -EINVAL;
+		return rc;
 	}
-	dev = device_get_binding(argv[args_indx.port]);
 
-	if (dev != NULL) {
-		index = (uint8_t)atoi(argv[2]);
-		shell_fprintf(sh, SHELL_NORMAL, "Blinking port %s index %d.", argv[1], index);
-		shell_fprintf(sh, SHELL_NORMAL, " Hit any key to exit");
+	shell_fprintf(sh, SHELL_NORMAL, "Blinking port %s index %d.", argv[1], index);
+	shell_fprintf(sh, SHELL_NORMAL, " Hit any key to exit");
 
-		/* dummy read to clear any pending input */
-		(void)sh->iface->api->read(sh->iface, &data, sizeof(data), &count);
+	/* dummy read to clear any pending input */
+	(void)sh->iface->api->read(sh->iface, &data, sizeof(data), &count);
 
-		while (true) {
-			(void)sh->iface->api->read(sh->iface, &data, sizeof(data), &count);
-			if (count != 0) {
-				break;
-			}
-			gpio_pin_set(dev, index, value);
-			value = !value;
-			k_msleep(SLEEP_TIME_MS);
+	while (true) {
+		(void)sh->iface->api->read(sh->iface, &data, sizeof(data), &count);
+		if (count != 0) {
+			break;
 		}
-
-		shell_fprintf(sh, SHELL_NORMAL, "\n");
+		gpio_pin_set(dev, index, value);
+		value = !value;
+		k_msleep(SLEEP_TIME_MS);
 	}
 
+	shell_fprintf(sh, SHELL_NORMAL, "\n");
+
 	return 0;
 }
 
-SHELL_STATIC_SUBCMD_SET_CREATE(sub_gpio,
-			       SHELL_CMD_ARG(conf, NULL, "Configure GPIO", cmd_gpio_conf, 4, 0),
-			       SHELL_CMD_ARG(get, NULL, "Get GPIO value", cmd_gpio_get, 3, 0),
-			       SHELL_CMD_ARG(set, NULL, "Set GPIO", cmd_gpio_set, 4, 0),
-			       SHELL_CMD_ARG(blink, NULL, "Blink GPIO", cmd_gpio_blink, 3, 0),
-			       SHELL_SUBCMD_SET_END /* Array terminated. */
-			       );
+static void device_name_get(size_t idx, struct shell_static_entry *entry)
+{
+	const struct device *dev = shell_device_lookup(idx, NULL);
+
+	entry->syntax = (dev != NULL) ? dev->name : NULL;
+	entry->handler = NULL;
+	entry->help = NULL;
+	entry->subcmd = NULL;
+}
+
+SHELL_DYNAMIC_CMD_CREATE(dsub_device_name, device_name_get);
+
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	sub_gpio,
+	SHELL_CMD_ARG(conf, &dsub_device_name,
+		     "Configure GPIO\n"
+		     "Usage: conf <device> <pin> <mode>\n"
+		     "The mode argument can be an integer, or one of:\n"
```

Comment:
I think it's better to rephrase it to emphasize default "premade" values and custom nature of integer, something like this:

Mode argument accepts these values:
1. in - Input (floating)
2. inu - Input (with pull-up)
3. ind - Input (with pull-down)
4. out - Output (push-pull)
5. <any integer> - Custom mode value passed directly to gpio_pin_configure() flags argument.